### PR TITLE
yaml-cpp: add version 0.7.0 + cleanup + modernize

### PR DESCRIPTION
**Pull request by theirix**
Specify library name and version:  **yaml-cpp/0.7.0**

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by jgsogo — APPROVED** [review]

File: recipes/yaml-cpp/all/patches/0001-install-0.7.0.patch [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
++++ b/CMakeLists.txt
+@@ -24,9 +24,8 @@ option(YAML_BUILD_SHARED_LIBS "Build yaml-cpp shared library" ${BUILD_SHARED_LIB
+ cmake_dependent_option(YAML_CPP_BUILD_TESTS
+   "Enable yaml-cpp tests" ON
+   "BUILD_TESTING;CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
+-cmake_dependent_option(YAML_CPP_INSTALL
+-  "Enable generation of yaml-cpp install targets" ON
+-  "CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
++option(YAML_CPP_INSTALL
++  "Enable generation of yaml-cpp install targets" ON)
```
Comment (madebr) [inline recipes/yaml-cpp/all/patches/0001-install-0.7.0.patch:14]: With this change (or something in the same spirit), this patch could be upstreamed:
```suggestion
@@ -24,9 +24,12 @@ option(YAML_BUILD_SHARED_LIBS "Build yaml-cpp shared library" ${BUILD_SHARED_LIB
 cmake_dependent_option(YAML_CPP_BUILD_TESTS
   "Enable yaml-cpp tests" ON
   "BUILD_TESTING;CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
-cmake_dependent_option(YAML_CPP_INSTALL
-  "Enable generation of yaml-cpp install targets" ON
-  "CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
+set(YAML_MAIN_PROJECT OFF)
+if(CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
+  set(YAML_MAIN_PROJECT ON)
+endif()
+option(YAML_CPP_INSTALL
+  "Enable generation of yaml-cpp install targets" ${YAML_MAIN_PROJECT})
```

**Review by madebr — DISMISSED** [review]

File: recipes/yaml-cpp/all/patches/0001-install-0.7.0.patch (hunk 1 above)
Comment (theirix) [reply recipes/yaml-cpp/all/patches/0001-install-0.7.0.patch:14]: It is already reported but not merged yet - see patch header
https://github.com/conan-io/conan-center-index/pull/6268/files/14615cf8fefee13d5d6350e3bb8765a45ab6f42a#diff-552386d9e4a343fcd90d5921ad7d29ff59efa772289097df48da091880df7b93R1

Comment (madebr) [reply recipes/yaml-cpp/all/patches/0001-install-0.7.0.patch:14]: Oops. Thanks!

File: recipes/yaml-cpp/all/conanfile.py [hunk 2]
```diff
@@ -60,10 +68,13 @@ def package(self):
         tools.rmdir(os.path.join(self.package_folder, 'lib', 'cmake'))
         tools.rmdir(os.path.join(self.package_folder, 'CMake'))
         tools.rmdir(os.path.join(self.package_folder, 'lib', 'pkgconfig'))
+        tools.rmdir(os.path.join(self.package_folder, 'share'))
 
     def package_info(self):
-        self.cpp_info.libs = tools.collect_libs(self)
-        if self.settings.os == "Linux":
+        self.cpp_info.libs = ["yaml-cpp"]
```
Comment (madebr) [inline recipes/yaml-cpp/all/conanfile.py:74]: Debug libraries have a `d` suffix.

Comment (theirix) [reply recipes/yaml-cpp/all/conanfile.py]: that's right. cannot understand where is 'md' from (`libyaml-cppmd`). reverted to `collect_libs` :(

Comment (madebr) [reply recipes/yaml-cpp/all/conanfile.py]: The suffix/postfix is applied here:
https://github.com/jbeder/yaml-cpp/blob/b591d8ae2ad1ff373273c3e05973adf6c46abfa8/CMakeLists.txt#L128

**Comment by theirix** [comment]
Configuration with DLL and debug MT (MTd) fails with an access violation for Conan build and for a pure local test. Excluded this configuration.

Not sure if a multithreaded library with DLL is a valid combination at all even if the DLL and the `test_package.exe` use the same runtime (library `libcmtd.lib`). Does anyone have any thoughts?

**Comment by madebr** [comment]
`MT` and DLLs can work if you never share file descriptors.
But the safest is of course to just disallow it. It's very error prone when your API is not adapted to it.

File: recipes/yaml-cpp/all/conanfile.py [hunk 3]
```diff
@@ -23,18 +26,27 @@ def _source_subfolder(self):
...
             del self.options.fPIC
 
     def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
         if self.settings.compiler.cppstd:
             tools.check_min_cppstd(self, "11")
 
+    def validate(self):
+        if self.settings.compiler == "Visual Studio" and self.options.shared and self.settings.compiler.runtime == "MTd":
```
Comment (madebr) [inline recipes/yaml-cpp/all/conanfile.py:43]: Usually we check for a static runtime with `"MT" in self settings.compiler.runtime`

Comment (theirix) [reply recipes/yaml-cpp/all/conanfile.py:43]: Thanks for the clarification.

> Usually we check for a static runtime with `"MT" in self settings.compiler.runtime`

Yes, I have seen this approach. I tried MT and it worked fine locally and on CI (as opposed to MTd). Probably better to ban both MT and MTd.

Comment (madebr) [reply recipes/yaml-cpp/all/conanfile.py:43]: If it fails with `MTd` but succeeds with `MT`, then it feels like it's a bug or we're entering undefined behavior land.
I'm thinking it's the latter so would just disable shared and MT.

The C++ allocator will not be shared between exe and sell, which will cause problems.

Comment (madebr) [reply recipes/yaml-cpp/all/conanfile.py:43]: If it fails with `MTd` but succeeds with `MT`, then it feels like it's a bug or we're entering undefined behavior land.
I'm thinking it's the latter so would just disable shared and MT.

The C++ allocator will not be shared between exe and dll, which will cause problems.

Comment (theirix) [reply recipes/yaml-cpp/all/conanfile.py]: Yes, better safe than sorry. Disabled it.

File: recipes/yaml-cpp/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -2,8 +2,10 @@ cmake_minimum_required(VERSION 3.1)
...
 
 include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
-conan_basic_setup()
+conan_basic_setup(TARGETS)
+
+find_package(yaml-cpp REQUIRED CONFIG)
 
 add_executable(${PROJECT_NAME} test_package.cpp)
-target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
+target_link_libraries(${PROJECT_NAME} yaml-cpp::yaml-cpp)
```
Comment (madebr) [inline recipes/yaml-cpp/all/test_package/CMakeLists.txt:10]: You're creating a `yaml-cpp` aliased target in the main recipe.
```suggestion
target_link_libraries(${PROJECT_NAME} yaml-cpp)
```

**Review by madebr — COMMENTED** [review]
Also, while you're modernizing this recipe.
If you feel up to it, can you replace all usages of single quotes with double quotes?

**Review by madebr — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**